src/lib.rs

```diff
@@ -1,42 +1,64 @@
-use serde::Deserialize;
 use url::Url;
 
 mod clients;
+mod errors;
 pub mod types;
 
 pub use clients::*;
+pub use errors::*;
 
-// TODO: more info, i.e. api path etc, for error.
-#[derive(thiserror::Error, Debug)]
-pub enum Error {
-    #[error("error making request request")]
-    RequestError(#[from] reqwest::Error),
-    #[error("error")]
-    Ngrok(NgrokError),
-    #[error("unknown error returned: {0}")]
-    UnknownError(String),
-}
+/// `ClientBuilder` is a builder for [Client]
+pub struct ClientBuilder {
+    // required options
+    api_key: String,
 
-#[derive(Debug, Deserialize)]
-pub struct NgrokError {
-    // TODO: parse out the code, we really shouldn't be using a string for this.
-    pub error_code: String,
-    pub msg: String,
+    // optional
+    api_url: Option<Url>,
+    reqwest_client: Option<reqwest::Client>,
 }
 
-/// Configuration options for constructing a [Client].
-#[derive(Clone, Debug)]
-pub struct ClientConfig {
-    /// The ngrok API Key to authenticate with. See the [API documentat](https://ngrok.com/docs/api) for more information on creating an ngrok API key.
-    pub api_key: String,
-    /// The URL to connect to. By default, `https://api.ngrok.com`.
-    pub api_url: Option<Url>,
+/// `ClientBuilder` allows customizing various options for the [Client]
+impl ClientBuilder {
+    /// Construct a builder to customize and then build the client.
+    pub fn new(api_key: String) -> ClientBuilder {
+        ClientBuilder {
+            api_key,
+            api_url: None,
+            reqwest_client: None,
+        }
+    }
+
+    /// Set an API url to connect to. By default, the public ngrok API will be used.
+    pub fn api_url(&mut self, api_url: Url) -> &mut Self {
+        self.api_url = Some(api_url);
+        self
+    }
+
+    /// Set a custom reqwest client to use for http requests to the ngrok API.
+    pub fn reqwest_client(&mut self, client: reqwest::Client) -> &mut Self {
+        self.reqwest_client = Some(client);
+        self
+    }
```

Comment:
Ah, the age-old debate. Builder-by-ownership or builder-by-mutable-reference.

I prefer taking/returning owned `Self`s, but I won't block on it :stuck_out_tongue:

Reply:
I picked this based on https://docs.rs/derive_builder/latest/derive_builder/#builder-patterns

My preference is mutable because I can imagine code like:

```rust
let mut builder = ClientBuilder::new(key);
if let Some(url) = env::var("NGROK_API_URL") {
   builder.api_url(url.parse()?);
}
let client = builder.build();
```

Admittedly, it's unlikely most people will want that for api_url, but we may add some other flags (debug or such) where it makes more sense, and I think that looks a bit cleaner than the owned form (`builder = builder.api_url`).